pkg/segment/aggregations/segaggs.go

```diff
@@ -2114,46 +2114,80 @@ func getAggregationResultFieldValues(fieldToValue map[string]segutils.CValueEncl
 
 func performTransactionCommandRequest(nodeResult *structs.NodeResult, aggs *structs.QueryAggregators, recs map[string]map[string]interface{}, finalCols map[string]bool, numTotalSegments uint64, finishesSegment bool) {
 
-	if finishesSegment {
-		nodeResult.RecsAggsProcessedSegments++
-	}
-
 	if recs != nil {
 
 		if nodeResult.TransactionEventRecords == nil {
 			nodeResult.TransactionEventRecords = make(map[string]map[string]interface{})
 		}
 
-		for k := range recs {
-			nodeResult.TransactionEventRecords[k] = recs[k]
-			delete(recs, k)
+		if nodeResult.TransactionsProcessed == nil {
+			nodeResult.TransactionsProcessed = make(map[string]map[string]interface{}, 0)
 		}
 
-		if nodeResult.RecsAggsProcessedSegments < numTotalSegments {
-			return
+		if aggs.TransactionArguments.SortedRecordsSlice == nil {
+			aggs.TransactionArguments.SortedRecordsSlice = make([]map[string]interface{}, 0)
 		}
 
-		records, cols, err := processTransactionsOnRecords(nodeResult.TransactionEventRecords, nil, aggs.TransactionArguments)
-		if err != nil {
-			log.Errorf("performTransactionCommandRequest: %v", err)
-			return
+		for k, v := range recs {
+			nodeResult.TransactionEventRecords[k] = recs[k]
+			aggs.TransactionArguments.SortedRecordsSlice = append(aggs.TransactionArguments.SortedRecordsSlice, map[string]interface{}{"key": k, "timestamp": v["timestamp"]})
+			delete(recs, k)
 		}
 
-		nodeResult.TransactionEventRecords = nil // Clear the transaction records
-		nodeResult.TransactionEventRecords = make(map[string]map[string]interface{})
-		nodeResult.TransactionEventRecords["CHECK_NEXT_AGG"] = make(map[string]interface{})
+		var cols []string
+		var err error
 
-		for i, record := range records {
-			recs[i] = record
-		}
+		if finishesSegment {
+			nodeResult.RecsAggsProcessedSegments++
+
+			// Sort the records by timestamp. The records in the segment may not be sorted. We need to sort them before processing.
+			// This method also assumes that all records in the segment will come before the records in the next segment(Segments are Sorted).
+			sort.Slice(aggs.TransactionArguments.SortedRecordsSlice, func(i, j int) bool {
+				return aggs.TransactionArguments.SortedRecordsSlice[i]["timestamp"].(uint64) < aggs.TransactionArguments.SortedRecordsSlice[j]["timestamp"].(uint64)
+			})
+
+			cols, err = processTransactionsOnRecords(nodeResult.TransactionEventRecords, nodeResult.TransactionsProcessed, nil, aggs.TransactionArguments, nodeResult.RecsAggsProcessedSegments == numTotalSegments)
+			if err != nil {
+				log.Errorf("performTransactionCommandRequest: %v", err)
+				return
+			}
+
+			nodeResult.TransactionEventRecords = nil // Clear the transaction records. Release the memory.
+			nodeResult.TransactionEventRecords = make(map[string]map[string]interface{})
```

Comment:
Don't you have to keep all the records with a timestamp that's after the last record that got into a transaction, because those records might be in part of a transaction that gets finished by records in the next segment you process? Or are those records already being stored somewhere else, like inside a partial transaction?

Reply:
Yes, they are being stored inside a partial transaction. They are being stored in `aggs.TransactionArguments.OpenTransactionEvents`. 
So I created two fields for `TransactionArguments`; 1. To store the events/records. 2. To store the state of the Transaction.

The `nodeResult.TransactionEventRecords` is only being used to fetch all the records per segment.